poly/sharing_test.go

```diff
@@ -0,0 +1,544 @@
+	*PriPoly, *PriPoly) {
+	return producePriPoly(g, k, s), producePriPoly(g, k, s),
+		producePriPoly(g, k, s)
+}
+
+func producePriShares(g abstract.Group, k, n int, s abstract.Secret) *PriShares {
+
+	testPoly := producePriPoly(g, k, s)
+	return new(PriShares).Split(testPoly, n)
+}
+
+func producePubPoly(g abstract.Group, k, n int, s abstract.Secret,
+	p abstract.Point) *PubPoly {
+
+	testPriPoly := producePriPoly(g, k, s)
+	testPubPoly := new(PubPoly)
+	testPubPoly.Init(g, n, p)
+	return testPubPoly.Commit(testPriPoly, p)
+}
+
+func producePubShares(g abstract.Group, k, n, t int, s abstract.Secret,
+	p abstract.Point) *PubShares {
+
+	testPubPoly := producePubPoly(g, k, n, s, p)
+	return new(PubShares).Split(testPubPoly, t)
+}
+
+/* Test Functions */
+
+// Test that the Pick function creates unique polynomials and unique secrets.
+func TestPriPolyPick_UniqueShares(t *testing.T) {
+	testPoly1, testPoly2, testPoly3 := produceThreePriPoly(group, k, nil)
+	if testPoly1.Equal(testPoly2) || testPoly1.Equal(testPoly3) ||
+		testPoly2.Equal(testPoly3) {
+		t.Error("Failed to create unique polynomials.")
+	}
+	if testPoly1.Secret().Equal(testPoly2.Secret()) ||
+		testPoly1.Secret().Equal(testPoly3.Secret()) ||
+		testPoly2.Secret().Equal(testPoly3.Secret()) {
+		t.Error("Failed to create unique secrets.")
+	}
+}
+
+// Test polynomials that are based on common secrets. Verify that
+// unique polynomials are created but that the base secrets are all the same.
+func TestPriPolyPick_CommonShares(t *testing.T) {
+	testPoly1, testPoly2, testPoly3 := produceThreePriPoly(group, k, secret)
+	if testPoly1.Equal(testPoly2) || testPoly1.Equal(testPoly3) ||
+		testPoly2.Equal(testPoly3) {
+		t.Error("Failed to create unique polynomials.")
+	}
+	if !testPoly1.Secret().Equal(testPoly2.Secret()) ||
+		!testPoly1.Secret().Equal(testPoly3.Secret()) ||
+		!testPoly2.Secret().Equal(testPoly3.Secret()) {
+		t.Error("Polynomials are expected to have the same secret.")
+	}
+}
+
+// Verify that the secret function works. The function should return the same
+// secret it was initialized with.
+func TestPriPolySecret(t *testing.T) {
+	testPoly := producePriPoly(group, k, secret)
+	if !secret.Equal(testPoly.Secret()) {
+		t.Error("This secret differs from the one given to it.")
+	}
+}
+
+// Verify that Equal returns true for two polynomials that are the same
+func TestPriPolyEqual_Same(t *testing.T) {
+	testPoly := producePriPoly(group, k, secret)
+	testPolyCopy := testPoly
+	if !testPoly.Equal(testPolyCopy) {
+		t.Error("Polynomials are expected to be equal.")
+	}
+}
+
+// Verify that Equal returns false for two polynomials that are diffferent
+func TestPriPolyEqual_Different(t *testing.T) {
+	testPoly1, testPoly2 := produceTwoPriPoly(group, k, secret)
+	if testPoly1.Equal(testPoly2) {
+		t.Error("Polynomials are expected to be different.")
+	}
+}
+
+// Verify that Equal panics if the polynomials are of different degrees.
+func TestPriPolyEqual_Error1(t *testing.T) {
```

Comment:
Convert the _Error1 and _Error2 functions to a single function as discussed above as well, using an internal helper function to handle the 'defer' clause for each case.  For example, something like:

func TestPriPolyEqual_Error2(t *testing.T) {
        test := func(p1, p2 *PriPoly) {
   	      defer deferTest(t, "The Equal method should have panicked.")
              p1.Equal(p2)
        }

        // different degrees
	testPoly1 := producePriPoly(group, k, secret)
	testPoly2 := producePriPoly(group, k+10, secret)
	test(testPoly1, testPoly2)

        // different groups
	testPoly1 := producePriPoly(group, k, secret)
	testPoly2 := producePriPoly(altGroup, k, altSecret)
	test(testPoly1, testPoly2)
}